californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
4. only the first response is accessible through MessageObserver.onResponse()
5. only the next notification are available on NotificationListener.

[Here](https://github.com/eclipse/californium/pull/16#commitcomment-16503803) is the main explanation of this.